flat_mutation_reader.cc

```diff
@@ -971,6 +971,105 @@ make_flat_mutation_reader_from_fragments(schema_ptr schema, reader_permit permit
     return make_flat_mutation_reader_from_fragments(std::move(schema), std::move(permit), std::move(filtered), pr);
 }
 
+flat_mutation_reader
+make_slicing_filtering_reader(flat_mutation_reader rd, const dht::partition_range& pr, const query::partition_slice& slice) {
+    class reader : public flat_mutation_reader::impl {
+        flat_mutation_reader _rd;
+        const dht::partition_range* _pr;
+        const query::partition_slice* _slice;
+        dht::ring_position_comparator _cmp;
+        std::optional<clustering_ranges_walker> _ranges_walker;
+        std::optional<range_tombstone_splitter> _splitter;
+
+    public:
+        reader(flat_mutation_reader rd, const dht::partition_range& pr, const query::partition_slice& slice)
+            : flat_mutation_reader::impl(rd.schema(), rd.permit())
+            , _rd(std::move(rd))
+            , _pr(&pr)
+            , _slice(&slice)
+            , _cmp(*_schema) {
+        }
+
+        virtual future<> fill_buffer(db::timeout_clock::time_point timeout) override {
+            const auto consume_fn = [this] (mutation_fragment mf) {
+                push_mutation_fragment(std::move(mf));
+            };
+
+            return do_until([this] { return is_buffer_full() || is_end_of_stream(); }, [this, consume_fn, timeout] {
+                return _rd.fill_buffer(timeout).then([this, consume_fn] {
+                    while (!_rd.is_buffer_empty()) {
+                        auto mf = _rd.pop_mutation_fragment();
+                        switch (mf.mutation_fragment_kind()) {
```

Comment:
How about writing this as a coroutine? do_until becomes `while`, and .then becomes `co_await`. It's not strictly required but it's easy and improves readability.

Reply:
Done